### ISSUE TITLE
Improve begin support

### ISSUE THREAD

**Issue by luisggpina** [daniel-grumberg/varan-div#4]
`begin` currently supports a single syscall, right?

After looking at how Valgrind starts executing the program after executing the loader, I would like to write something like this for `begin`:

```
open() = fd,
*,
close(fd),
mmap(_, _, _, _, -1, 0),
mmap(_, _, _, _, -1, 0),
mprotect(_, _, PROT_READ) * ,
munmap(_, _)
```

This means
- Match an `open` and save the result as `fd`
- Match any system call until a `close` is issued with the same `fd` as the first matched open
- Match two `mmap`s with `-1,0` as the last two arguments
- Match any number of `mprotect` with `PROT_READ` as the last argument
- Match one `munmap`

The idea is that Varan would look for this.  If the syscall sequence diverges, given that this is the `begin`, it would simply update its state to go back to the beginning and try to match this sequence again with the rest of the system calls that the follower will issue.

Of course, we can move all the conditions on the arguments to a `{ ... }` block at first, to allow a quick implementation.  But I think this syntax looks neat and tidy.

Thoughts?

**Comment by daniel-grumberg** [comment]
That is impossible with to achieve with the current model. What I mean by that is that the DSL never has any knowledge of the return values of the system calls. There are 2 ways to handle this IMHO:

- We execute the system calls within the DSL, which I doubt is feasible
- We pass in a dictionary to the matching function, which might look a bit clunky but would keep the separation between the DSL and actual Varan.

So that is a major problem at the moment.

**Comment by luisggpina** [comment]
Yes, you're right.  Executing syscalls is Varan's part.  I'll think about this part on the background.

What about the rest, ignoring the open-close stuff?  In particular, the star operator after a syscall?

**Comment by daniel-grumberg** [comment]
Any number of a system call/system calls is more complicated but is do-able with a bit of work. The issue here is to know when to swap out the function pointer, but I am pretty sure I can make work if I introduce an extra global variable within  the `.c` file.

**Comment by luisggpina** [comment]
Ok, that sounds good.  After more careful thinking, I believe we can get away without the open-close part to detect when Valgrind starts executing a program.

**Comment by andronat** [comment]
Can't this issue be solved with a global state variable as we discussed at some point? The automaton can return with a special value directing Varan to store the return value of the syscall and then we can access the global value in `{ ... }`.

Also, I remember that we also discussed `*` to be added but there were some concerns. Does anyone remember what was the problem? Because I would like to have `*`.
